proposals/052-cryptography-leg-1.md

```diff
@@ -0,0 +1,329 @@
+
+- Insufficient abstraction
+- Maintenance of handwritten C
+- Bindings to an insufficiently general library
+- Burdensome dependencies
+- Dead project
+- Lack of post-quantum algorithms
+
+However, there is also a lot to be learned from the ways these prior efforts *were* successful. This proposal will assess these existing efforts in an attempt to consolidate their learnings and functionality.
+
+## An example of a successful ecosystem
+
+The health of the Haskell server ecosystem may be used as an example of a successful community-driven niche, and we can aspire to bring that same quality and health to the Haskell cryptography ecosystem. There are many pertinent parallels in terms of managing performance, complexity, and safety that we can learn from:
+
+- Low-level packages like `wai` supporting multiple backends
+- Backends range from simple like `scotty` to complex like `servant`
+- Frontends like `blaze-html` and `lucid`
+- Libraries can take advantage of type system to provide good abstractions
+- Flexible enough abstractions to explore the problem space with multiple solutions
+- Appropriate licensing
+
+`servant` (and the `wai` ecosystem in general) is a success story and an example of higher-order programming making something safer and easier, something that we can aspire to do with cryptography.
+
+## Technical Content
+
+This proposal is for the contruction of a suite of cryptography libraries, ranging from low-level bindings to specific algorithms, to higher-level abstractions and typeclasses.
+
+This leg of the proposal focuses on bindings to the `botan` cryptography, in the form of the following libraries:
+
+### botan-bindings
+
+The `botan-bindings` library will contain raw bindings and is an almost direct, 1-1 translation of the C API into Haskell FFI calls. As such, it exposes and operates over C FFI types, requires buffer and pointer and pointer management, and returns error codes. This library only exposes FFI calls and constants, and is suitable for building your own abstraction over Botan.
+
+### botan-low
+
+The `botan-low` library will contain low-level bindings which wrap the FFI calls into monadic IO actions. This library handles the translation between buffers and ByteStrings, and throws exceptions in the case of errors, but is otherwise a fairly faithful translation of the interface. Thus this library is suitable for use, but is not recommended for day-to-day programming due to statefulness and a lack of type safety.
+
+### botan
+
+The `botan` library will contain high-level bindings which hides the stateful IO and exposes appropriately idiomatic and pure interfaces. It will provide data types for algorithms, enumerations for constants, convenience methods for key and nonce generation, incremental processing of lazy bytestrings, and otherwise bring significant safety and ergonomics over `botan-low` and `botan-bindings`.
```

Comment:
```suggestion
The `botan` library will contain high-level bindings which hide the stateful IO and expose appropriately idiomatic  pure interfaces. It will provide data types for algorithms, enumerations for constants, convenience methods for key and nonce generation, incremental processing of lazy bytestrings, and otherwise bring significant safety and ergonomics over `botan-low` and `botan-bindings`.
```